Issue warning if renaming branch with a remote tracking branch.

essiene opened the pull request:
Fixed #1457 

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
> N/A
- [ ] I have updated the config schema (cli/src/config-schema.json)
> N/A
- [X] I have added tests to cover my changes

martinvonz commented on cli/src/commands/branch.rs:
```diff
@@ -349,6 +349,21 @@ fn cmd_branch_rename(
+            &StringPattern::exact(old_branch),
+            &StringPattern::everything(),
+        )
+        .count()
+        > 0
```
This can be written `.next().is_some()`, which will terminate early instead of exhausting the iterator, but that shouldn't make a difference in practice in this case. Just a good pattern to keep in mind.

essiene replied:
Thanks. Noted the pattern. And Clippy helpfully suggests I can replace .filter().next() with .find()

essiene replied:
...aaaand Clippy suggests any() which takes a predicate anyway.

essiene commented on cli/src/commands/branch.rs:
```diff
@@ -349,6 +349,21 @@ fn cmd_branch_rename(
+            &StringPattern::exact(old_branch),
+            &StringPattern::everything(),
+        )
+        .count()
```
> insert `.filter(|(_, remote_ref)| remote_ref.is_tracking())` to exclude non-tracking remote branches.

Done.
> 
> (Unrelated to this, maybe we'll need to rename `is_tracking()` to `is_tracked()` per #2772)

I lol'd at the number of remotes and tracking in that thread ;-) Let me know if it's worth following up with a one-time cleanup as a different issue.

essiene commented on cli/tests/test_branch_command.rs:
Please help me check the approach to the testing here, if there is anything I could potentially simplify? It works, but maybe unweildly?

martinvonz commented on cli/src/commands/branch.rs:
```diff
@@ -350,6 +350,20 @@ fn cmd_branch_rename(
+    {
+        writeln!(
+            ui.warning(),
+            "warning: Branch: {}, has remote branches which will not be renamed",
```
I think it looks better without the ":" after "Branch"

essiene replied:
Thanks. Fixed.

martinvonz commented on cli/tests/test_branch_command.rs:
```diff
@@ -195,21 +195,54 @@ fn test_branch_rename() {
+
+    test_env.jj_cmd_ok(&repo_path, &["new"]);
+    test_env.jj_cmd_ok(&repo_path, &["describe", "-m=commit-2"]);
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["branch", "create", "branchwithtrackingremote"],
+    );
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["git", "push", "--branch=branchwithtrackingremote"],
+    );
```
It might be enough to do `jj git fetch` here and get the `master` branch from the remote, then `jj branch rename master renamed-master`

essiene replied:
So it looks like just adding a binary blob to the git bare repo is not enough to create the master branch that can be fetched. i.e. I would need to call repo.commit(...) which has a bunch of setup which ended up being more unweidly than the current version. I could be doing it wrong, but when I wrote a test and tried to 'jj git fetch' followed by 'jj git branch list' and I could not find master. 

So I took the lazy way out :-D

I went back to creating and pushing a branch, but I've renamed the branches in the test such that it makes a lot more sense to read through the lines. I think random names like 'foo' and 'bar' added to my initial dislike.

Let me know what you think.

martinvonz replied:
Makes sense.

essiene replied:
Thanks. Resolving.

yuja commented on cli/src/commands/branch.rs:
```diff
@@ -350,6 +350,20 @@ fn cmd_branch_rename(
+    {
+        writeln!(
+            ui.warning(),
+            "warning: Branch: {}, has remote branches which will not be renamed",
```
nit: excessive comma?
and the parameter can be inlined as `"Branch {old_branch} ..."`

essiene replied:
Fixed.

ilyagr reviewed (commented):
Please fix a confusing typo in PR and commit title. "renaming", not "remaining".

essiene commented:
> Please fix a confusing typo in PR and commit title. "renaming", not "remaining".

Whoops! Thanks. Fixed.

martinvonz commented:
> > Please fix a confusing typo in PR and commit title. "renaming", not "remaining".
> 
> Whoops! Thanks. Fixed.

Still there in the commit description.

essiene commented:
> > > Please fix a confusing typo in PR and commit title. "renaming", not "remaining".
> > 
> > 
> > Whoops! Thanks. Fixed.
> 
> Still there in the commit description.

doh! fixed for real this time!

martinvonz commented on cli/tests/snapshots/test_branch_command__.snap:
Please delete.

I assume this was created by `insta`. Do you know how? I wonder if we should add `*.snap` to our `.gitignore` file.

Oh, I see you sent #2787 for `*.snap.new`. Do we need both?

essiene replied:
..and I missed this. Yeah. I'll just update #2787 to simply do *.snap.*

essiene replied:
Thanks for the catch. Deleted the extra file.

Also update to `*.snap*` in #2787

martinvonz reviewed (approved):
Thanks!
